[1,2단계 - 체스] 냥인(최가희) 미션 제출합니다.

cutehumanS2 opened the pull request:
오리, 안녕하세요. 6기 냥인입니다.

리뷰 잘 부탁드립니다.◠‿◠

cutehumanS2 committed:
docs: 구현 기능 목록 작성

cutehumanS2 committed:
feat: 초기 체스판 생성

cutehumanS2 committed:
feat: 초기 보드 출력 기능

cutehumanS2 committed:
feat: Rook, Knight에 대한 이동 가능 여부 판단 기능

- Piece 클래스를 추상화
- Piece 클래스를 구현한 기물 클래스 생성
- PieceName -> PieceType로 네이밍 변경
- Rank 클래스를 record -> enum으로 변경

cutehumanS2 committed:
feat: King, Queen, Bishop, Pawn에 대한 이동 가능 여부 판단 기능

cutehumanS2 committed:
feat: 목적지까지의 경로 생성 기능

cutehumanS2 committed:
feat: Pawn을 제외한 모든 기물들의 이동 가능 여부 판단 기능

cutehumanS2 committed:
feat: 기물 이동 기능

cutehumanS2 committed:
feat: 게임 진행 기능

- 나이트는 이동 경로에 장애물이 있는지 판별하는 로직 제외
- 예외 메시지 변경

cutehumanS2 committed:
refactor: 포지션 관련 책임 포지션 객체로 위임

Co-Authored-By: Gyeongho Yang <chchch1213@naver.com>

cutehumanS2 committed:
refactor: Position -> Square 네이밍 변경

Co-Authored-By: Gyeongho Yang <chchch1213@naver.com>

cutehumanS2 committed:
refactor: 패키지 구조 변경

Co-Authored-By: Gyeongho Yang <chchch1213@naver.com>

cutehumanS2 committed:
refactor: 상수화 및 상수명 변경

Co-Authored-By: Gyeongho Yang <chchch1213@naver.com>

cutehumanS2 committed:
feat: 턴 전환 기능 및 네이밍 변경

PieceColor -> Team

Co-Authored-By: Gyeongho Yang <chchch1213@naver.com>

cutehumanS2 committed:
docs: 프로젝트 구조 추가

jinyoungchoi95 commented on src/main/java/chess/domain/square/File.java:
```diff
@@ -0,0 +1,49 @@
+    g(7),
+    h(8);
+
+    static final String ERROR_NOT_EXIST_FILE = "은(는) 존재하지 않는 파일입니다.";
```
```suggestion
    private static final String ERROR_NOT_EXIST_FILE = "은(는) 존재하지 않는 파일입니다.";
```
자바에서는 protected와 default를 제공하지만 완전한 접근제어자가 아닌 결국 접근이 가능한 접근제어자입니다. 미션을 진행하면서 가능하면 public과 private으로만 설계할 수 없는지 고민해보세요.

테스트에서 사용되는 경우 테스트 결과값은 클래스에서 사용하는 상수를 그대로 사용하기보단 무엇을 검증하는지는 테스트에게 위임해보세요.

cutehumanS2 replied:
말씀해 주신 대로 가능하면 public과 private 접근 제어자만 사용하기 위해 
일단 기존 예외 메시지를 검증하는 코드는 지우고, 해당 예외 메시지 상수들도 모두 private로 변경해 주었습니다.

그런데 `무엇을 검증할지는 테스트에게 맡긴다` 의 의미가 잘 이해되지 않습니다.

만약, 아래와 같이 검증 메서드가 많고 모두 `IllegalArgumentException`을 발생시킨다고 가정했을 때,
```java
public void move(final Square source, final Square target, final PieceColor turn) {
        validateIsSameSquare(source, target);
        validateIsNonExistentPiece(source);

        Piece sourcePiece = findPieceBySquare(source);
        validateIsTurn(sourcePiece, turn);
        validateCanMove(source, target, sourcePiece);
        validateExistObstacleOnPath(source, target);
        pieces.remove(source);
        pieces.put(target, sourcePiece);
    }
```

예외가 정말로 발생하는지 테스트에게 맡기려면 아래 코드 정도로만 작성해도 충분한가요?
```java
assertThatThrownBy(() -> board.move(source, target, turn))       
    .isInstanceOf(IllegalArgumentException.class);
```

어느 검증 메서드에서 발생한 예외인지 예외 메시지를 검증을 통해 명확히 알고 싶다면 어떻게 하면 될까요??  테스트 메서드 안에서 예외 메시지 값을 하드 코딩 해주거나 상수로 또 정의하는 게 좋을까요?

jinyoungchoi95 replied:
assertion에서 검증하는 부분이 production에 의존적이지 않아야한다는 이야기였어요.
```java
assertThatCode(() -> File.findByValue("i"))
                .isInstanceOf(IllegalArgumentException.class)
                .hasMessage("은(는) 존재하지 않는 파일입니다.");
```

예외의 메시지를 테스트하는 경우 저희는 `ERROR_NOT_EXIST_FILE`가 들어가는 것을 기대하는 것이 아니라 `은(는) 존재하지 않는 파일입니다.`라는 문구가 나오는지 테스트하기를 기대해요. 즉, 테스트는 production의 값을 검증하기 때문에 production 코드로 영향이 가는 테스트를 작성하면 안되는거죠.

TDD를 예를들면 테스트를 먼저 작성했더라면 `ERROR_NOT_EXIST_FILE`를 사용하는 테스트가 나왔었을까요? 제 생각에는 `"은(는) 존재하지 않는 파일입니다."`라는 String값을 테스트에서 사용했을 것 같아요. 테스트 케이스는 운영의 코드에 상수가 들어간다는 사실을 모르고 있으니까요.

https://jojoldu.tistory.com/615 글을 한번 참고해보시면 도움이 될 것 같네요 😄

jinyoungchoi95 commented on src/main/java/chess/domain/square/File.java:
```diff
@@ -0,0 +1,49 @@
+        }
+    }
+
+    public File add(int value) {
```
테스트코드가 필요하지 않을까요? 해당 클래스뿐만아닌 전반적으로 각 클래스가 가지는 단위테스트가 부족하니 모두 추가해주세요.

cutehumanS2 replied:
전반적으로 단위 테스트를 추가하였습니다! 
그러나 테스트 데이터는 좀 더 추가해야 하는 상태입니다.😿

jinyoungchoi95 commented on src/main/java/chess/domain/square/File.java:
```diff
@@ -0,0 +1,49 @@
+
+    static final String ERROR_NOT_EXIST_FILE = "은(는) 존재하지 않는 파일입니다.";
+
+    private final int file;
```
`File.file`은 어색한 표현아닐까요? 숫자가 지니는 의미가 무엇인지 명확하게 네이밍지어주시면 좋을 것 같아요.

cutehumanS2 replied:
값이 지니는 의미가 명확해지도록 `file -> index`로 변경해 보았습니다.

jinyoungchoi95 commented on src/main/java/chess/domain/square/File.java:
```diff
@@ -0,0 +1,49 @@
+        return file - target.file;
+    }
+
+    public int get() {
```
여기도 단순히 `get()`은 무엇을 가져오는지 알기 힘들어보이네요. 축약된 네이밍들은 모두 무엇을 표현하는지 명확하게해주세요. 미션은 학습을 위한 것이기 때문에 최대한 명확하게 네이밍 지으려고 연습해보는 것도 중요해요. 과해도 좋으니 다른사람이 코드를 읽었을 때 이해하기 쉬운 네이밍이 무엇인가 항상 고민해보세요.

cutehumanS2 replied:
마찬가지로 `getIndex()`로 변경해 보았습니다.
다른 부분도 모두 명확하게 네이밍하려 노력해 보았습니다.😀

jinyoungchoi95 commented on src/main/java/chess/domain/square/File.java:
```diff
@@ -0,0 +1,49 @@
+    }
+
+    private static void validateExist(String input) {
+        if (Arrays.stream(values()).noneMatch(file -> file.name().equals(input))) {
```
객체지향 생활 체조 규칙4. 한 줄에 점은 하나만 찍는다.

가독성은 나를 위한 것이기도하지만 상대방을 위한 것이기도해요. 어떻게하면 더 잘 표현할 수 있을지 고민해보세요.

cutehumanS2 replied:
해당 규칙을 최대한 준수하도록 개선해 보았습니다.

jinyoungchoi95 commented on src/main/java/chess/domain/square/Square.java:
```diff
@@ -0,0 +1,66 @@
+
+public record Square(File file, Rank rank) {
+
+    public static Square from(String square) {
+        File file = File.from(square.substring(0, 1));
+        Rank rank = Rank.from(square.substring(1, 2));
+        return new Square(file, rank);
+    }
```
1. 생성자와 정적팩토리메소드의 차이는 무엇인가요? 관습적인 네이밍인 `from`은 결국 생성자에서 객체를 생성하는 것과 의미전달이 크게 달라보이지 않아요.
2. `square`는 화면에서 전달한 정보를 그대로 사용해주시는 것으로 보이는데요. 화면의 요구사항이 `a:1 b:1`로 변경되면 도메인 코드의 변경이 생길 것 같네요.

cutehumanS2 replied:
1. 정적 팩토리 메서드명을 관습적인 네이밍으로 지었을 때, ‘의미있는 이름을 지을 수 있다’는 장점을 누리기 어렵다는 의견에 동의합니다. `File`과 `Rank`의 `from`은 의미가 명확해지도록 `findByValue`로 변경해 주었습니다. `Square`의 경우 정적 팩토리 메서드를 사용해서 얻는 이점이 없어 생성자 방식으로 변경하였습니다.

2. move 명령어의 포맷이 변경될 경우 도메인 코드도 변경될 것을 고려하여, 사용자 입력값을 File, Rank, Square로 변환하는 책임을 도메인 외부로 위임하였습니다. 해당 작업을 view의 `Command`와 controller인 `ChessGame`에서 나눠 하고 있는데, 적절한 책임 분리인지, 그리고 변경에 덜 취약한 구조가 되었는지는 확신이 안 섭니다. 😿

jinyoungchoi95 replied:
> 해당 작업을 view의 Command와 controller인 ChessGame에서 나눠 하고 있는데, 적절한 책임 분리인지, 그리고 변경에 덜 취약한 구조가 되었는지는 확신이 안 섭니다. 😿

지금 정도면 충분하다고 생각해요. 개인적으로 `Application`, `ChessGame`과 같은 객체는 콘솔 게임실행으로 인해 더럽혀질 수 밖에 없기 때문에 너무 고민하지 않고 view, domain에 대한 의존을 어떻게 잘 끊을까를 고민해주시면 좋을 것 같아요. (반영하지 않으셔도 될만한 내용이지만 책임을 더 분리하고싶으시다면 입력된 move의 index 정보들을 parsing해주는 view용 객체가 있어도 될 것 같네요.)

jinyoungchoi95 commented on src/main/java/chess/domain/square/Square.java:
```diff
@@ -0,0 +1,66 @@
+    }
+
+    private Square addVector(int vectorFile, int vectorRank) {
+        return new Square(
+                this.file.add(vectorFile),
+                this.rank.add(vectorRank));
```
```suggestion
        return new Square(this.file.add(vectorFile), this.rank.add(vectorRank));
```
(반영하지 않으셔도 됩니다.) 한칸 내리는 것은 생각보다 많은 가독성 저해를 일으키니 가능하면 내리지 않는 것이 좋습니다. 현재 코드는 길이가 짧아 내릴 필요가 없어보이기도하구요 :)

cutehumanS2 replied:
오리의 의견에 동의합니다. 반영하였습니다.😸

jinyoungchoi95 commented on src/main/java/chess/domain/piece/Team.java:
```diff
@@ -0,0 +1,14 @@
+    WHITE;
+
+    public Team next() {
+        if (this.equals(BLACK)) {
```
enum은 동등성 비교와 동일성 비교 중 무엇을 사용하는 것이 좋을까요?

cutehumanS2 replied:
이 코멘트를 보고 Enum의 비교 방식에 대해 학습하게 되었습니다.
저는 개인적으로 동일성 비교를 사용하는 것이 편할 것 같아요.

Enum은 `equals()` 메서드를 제공하고 있는데, 이 메서드 내부에서도 `==` 비교하고 있다고 하더라고요. 그래서 `==`를 쓰면 동일성 비교를 하고 있다는 걸 명시적으로 나타낼 수 있으니, 가독성이 좋을 것 같아요. 그리고 동일성 비교를 하면 컴파일 타임에 미리 타입 호환성도 체크해주니 에러를 잡기 편할 것 같다는 생각이 듭니다.

깊게 공부하지는 못해서 제 생각은 이 정도로 말씀드릴 수 있는데, 오리의 생각은 어떠신가요??

jinyoungchoi95 replied:
네 냥인이 해주신 말씀이 맞아요 :) enum은 싱글톤이라는것정도만 더 알아가시면 좋을 것 같습니다.

jinyoungchoi95 commented on src/main/java/chess/domain/piece/Piece.java:
```diff
@@ -0,0 +1,25 @@
+
+public abstract class Piece {
+
+    // TODO: Type 필드 제거 후 Getter 추상화
```
TODO는 모두 지워주세요 :)

jinyoungchoi95 commented on src/main/java/chess/domain/piece/Piece.java:
```diff
@@ -0,0 +1,25 @@
+public abstract class Piece {
+
+    // TODO: Type 필드 제거 후 Getter 추상화
+    private final PieceType type;
```
상속을 사용하면서 `instanceOf`를 대신할 `PieceType`이 있는 것으로 보이는데요. 결국 `PieceType`을 알아야한다는 것 자체가 `Piece`로 추상화한 목적이 불분명하다는 의미 아닐까요?

상속과 조합을 둘다 사용한 것으로 보이는데 과연 적절한 상속으로 추상화를 하고 있는지 한번 고민해보시면 좋을 것 같아요.

cutehumanS2 replied:
오리의 의견에 동의합니다.
추상화를 잘 못하고 있다고 느껴서 구조를 바꿔 보았는데, 지금 구조도 좋은 구조라고는 확신을 못 하겠습니다.😹

jinyoungchoi95 commented on src/main/java/chess/domain/piece/Piece.java:
```diff
@@ -0,0 +1,25 @@
+
+    public abstract boolean canMove(Square source, Square target);
+
+    public PieceType getType() {
+        return type;
+    }
+
+    public Team getColor() {
+        return color;
+    }
```
abstract로 상속을 한다면 상속받은 하위 클래스에서 override하여 캡슐화를 무너뜨릴 수 있겠네요. 보완책은 없을까요?

cutehumanS2 replied:
하위 클래스에서 오버라이딩할 가능성을 고려하지 못했네요!
final 키워드를 사용한다면 보완할 수 있을 것 같아요.

jinyoungchoi95 commented on src/main/java/chess/domain/piece/Pawn.java:
```diff
@@ -0,0 +1,39 @@
+    }
+
+    private boolean isBackward(Square source, Square target) {
+        if (getColor() == Team.BLACK) {
```
getter를 지양하라는 것은 객체가 서로에게 메시지를 보내는 형태로 만들라는 이야기와 같아요. 미션 중에 getter는 출력의 경우에만 사용하고 도메인 로직에서 getter를 사용하는 코드를 최대한 제거해보시면 좋을 것 같아요.

cutehumanS2 replied:
최대한 getter를 사용하지 않도록 개선해 보았습니다.😸

jinyoungchoi95 commented on src/main/java/chess/domain/piece/Knight.java:
```diff
@@ -0,0 +1,21 @@
+
+    @Override
+    public boolean canMove(Square source, Square target) {
+        return (source.calculateFileDiff(target.file()) == SIDE_STEP
+                && source.calculateRankDiff(target.rank()) == STRAIGHT_STEP) ||
+                (source.calculateFileDiff(target.file()) == STRAIGHT_STEP
+                        && source.calculateRankDiff(target.rank()) == SIDE_STEP);
```
```suggestion
       if (source.calculateFileDiff(target.file()) == SIDE_STEP && source.calculateRankDiff(target.rank()) == STRAIGHT_STEP) {
            return true;
        }
        return source.calculateFileDiff(target.file()) == STRAIGHT_STEP && source.calculateRankDiff(target.rank()) == SIDE_STEP;
```

조건절은 복잡하면 복잡할수록 가독성이 떨어지니 if를 최대한 활용해보는 방법도 있습니다 😄

cutehumanS2 replied:
반영하였습니다. 좋은 방법 감사합니다. ㅎㅎ

```java
public boolean isStraightAndDiagonal(final Square target) {
        if (file.calculateDistance(target.file) == ONE_STEP
                && rank.calculateDistance(target.rank) == TWO_STEP) {
            return true;
        }
        return file.calculateDistance(target.file) == TWO_STEP
                && rank.calculateDistance(target.rank) == ONE_STEP;
    }
```

jinyoungchoi95 commented on src/main/java/Application.java:
```
RNBQKBNR
PPPPPPPP
........
........
........
........
pppppppp
rnbqkbnr

move a2 a4
RNBQKBNR
PPPPPPPP
........
........
p.......
........
.ppppppp
rnbqkbnr

move b7 b5
RNBQKBNR
P.PPPPPP
........
.P......
p.......
........
.ppppppp
rnbqkbnr

move a4 b5
[ERROR] 해당 위치로 기물을 이동할 수 없습니다.
```
a4 b5로 대각선말을 잡을 수 있어야하는 것 아닐까요?

cutehumanS2 replied:
해당 로직을 추가로 구현하였습니다.😸

jinyoungchoi95 commented on src/main/java/chess/domain/board/BoardFactory.java:
```diff
@@ -0,0 +1,52 @@
+
+public class BoardFactory {
+
+    public static final List<Function<Team, Piece>> PIECES_ARRANGEMENT = List.of(
```
👍

jinyoungchoi95 commented on src/test/java/chess/domain/board/BoardFactoryTest.java:
```diff
@@ -0,0 +1,63 @@
+    void createBoard() {
+        Board actual = BoardFactory.createBoard();
+
+        Map<Square, Piece> expected = new HashMap<>();
+        expected.put(new Square(File.a, Rank.EIGHT), new Rook(Team.BLACK));
```
`Map.of`를 사용해도 되지 않을까요?

cutehumanS2 replied:
`Map.of()`는 한번에 10개까지만 가능하다고 해서 `Map.ofEntries()`를 사용해 보았습니다!

jinyoungchoi95 commented on src/main/java/chess/domain/board/Board.java:
```diff
@@ -0,0 +1,60 @@
+    }
+
+    public void move(final Square source, final Square target) {
+        Piece sourcePiece = values.get(source);
+
+        if (isNotExistPiece(source)) {
+            throw new IllegalArgumentException(ERROR_NOT_EXIST_PIECE);
+        }
```
`values에서 Piece를 가져온 후 -> values에 Piece가 있는지 확인한다`

순서가 반대가 되어야하지 않을까요?

cutehumanS2 replied:
아래와 같이 순서를 변경하였습니다. 😺

```java
public void move(final Square source, final Square target, final PieceColor turn) {
        validateIsSameSquare(source, target);
        validateIsNonExistentPiece(source);

        Piece sourcePiece = findPieceBySquare(source);
        validateIsTurn(sourcePiece, turn);
        validateCanMove(source, target, sourcePiece);
        validateExistObstacleOnPath(source, target);
        pieces.remove(source);
        pieces.put(target, sourcePiece);
    }
```

jinyoungchoi95 commented on src/main/java/chess/domain/board/Board.java:
```diff
@@ -0,0 +1,60 @@
+        return path.stream().anyMatch(values::containsKey);
+    }
+
+    public boolean isNotExistPiece(Square square) {
+        return !values.containsKey(square);
+    }
```
냥인은 테스트를 위해서 public으로 열어두는 것에 대해서 어떻게 생각하시나요?

cutehumanS2 replied:
테스트를 위해 public으로 열어 두는 것에 반대하는 입장입니다.   

public으로 열어두면 테스트하기는 확실히 쉬워지겠지만, 테스트의 용이함보다는 외부에서 접근할 수 있어 캡슐화가 깨진다는 점을 더 우선시해야 한다고 봅니다. 

외부에서 접근할 수 있도록 의도한 메서드 외에는 private로 닫아 두었습니다. 😀

jinyoungchoi95 commented on src/main/java/chess/view/Command.java:
```diff
@@ -0,0 +1,24 @@
+
+    public static final String ERROR_INVALID_COMMAND = " 은(는) 존재하지 않는 명령어 입니다.";
+    private final String value;
+
+
```
```suggestion

```
👀

jinyoungchoi95 reviewed (changes requested):
안녕하세요 냥인 😄

체스 미션의 리뷰어를 맡은 오리입니다.

어느덧 레벨1의 마지막 미션이네요🙂 몇가지 코멘트 남겨두었으니 확인부탁드려요.

궁금하거나 고민이 되는 부분이 있으시다면 언제든 pr 코멘트 또는 dm으로 요청 부탁드립니다.
감사합니다 🙇‍♂️

cutehumanS2 committed:
refactor: 전체적인 구조 및 로직 변경

- Piece 추상 클래스 -> 일반 클래스로 전환, 각 기물의 이동 전략을 PieceType에서 관리하도록 변경
- Team -> PieceColor 네이밍 변경
- 에러 메시지의 접근 제어자를 default -> private로 변경
- File과 Rank 클래스의 필드명을 각각 index로 변경
- 기물의 이동 경로에 방해물이 존재하는지 확인하는 로직 변경

cutehumanS2 committed:
refactor: 불필요한 코드 제거

cutehumanS2 committed:
refactor: 게임 진행 로직 변경, 재입력 기능 제거

cutehumanS2 committed:
test: 단위 테스트 작성

cutehumanS2 committed:
feat: 폰이 색상에 따라 위 또는 아래로만 이동하도록 조건 추가

cutehumanS2 committed:
refactor: 정적 팩토리 메서드명 변경 및 생성자 호출 방식으로 변경

- File, Rank의 정적 팩토리 메서드명을 from -> findByValue로 변경
- Square의 생성 방식을 정적 팩토리 메서드 -> 생성자 호출 방식으로 변경

cutehumanS2 committed:
refactor: 테스트 코드에서 사용할 보드 초기화 방식 변경

HashMap 생성 후 put -> Map.ofEntries로 바로 생성

cutehumanS2 committed:
refactor: 비교 방식 변경(동등성 -> 동일성)

cutehumanS2 commented:
오리, 안녕하세요😸

전반적으로 로직에 구멍이 많아서 이것저것 변경을 하느라 리뷰 요청이 늦어졌습니다. 
지금도 개선해야 할 점이 많이 보이는데, 여기서 더 늦게 드리면 안 될 것 같아 일단 요청드립니다.

이번 리뷰도 잘 부탁드립니다.

### 주요 변경 사항

#### 1. Piece들 구조 변경

기존에는 Piece를 추상 클래스로 정의하고 나머지 기물들이 이를 상속받는 구조로 작성했었습니다.
그러나 상속 방식의 단점도 마음에 걸리고, 추상화를 잘 못하고 있는 것 같아 구조를 변경하였습니다.

현재는 Piece가 PieceType을 필드로 갖고 있고, 이 PieceType 안에서 각 기물의 이동 전략을 관리하고 있습니다.
```java
public enum PieceType {

    KING((source, target) -> (source.isStraight(target) || source.isDiagonal(target))
            && source.isWithinOneStep(target)),
    QUEEN((source, target) -> source.isStraight(target) || source.isDiagonal(target)),
    ROOK((source, target) -> source.isStraight(target)),
    BISHOP((source, target) -> source.isDiagonal(target)),
    KNIGHT((source, target) -> source.isStraightAndDiagonal(target)),
    PAWN((source, target) -> source.isOnlyForward(target) || source.isAttack(target));

    ...
```
그런데 이렇게 변경하고 나니 새로운 문제가 발생하네요. ^_^;;

Piece에 아래와 같은 코드가 생겨 버렸습니다.
폰은 색상에 따라 전진 방향을 체크해줘야 하니 아래처럼 보기 안 좋은 조건문이 생겨났습니다.
```java
public boolean canMove(final Square source, final Square target) {
        if (type == PieceType.PAWN) {
            return type.findMoveStrategy(source, target) && source.isNotBackward(target, color);
        }
        return type.findMoveStrategy(source, target);
    }
```
폰을 어떻게 처리해야 괜찮은 구조가 나올까요? 

Piece를 다시 추상 클래스로 만들고 폰만 Piece를 상속받게 해서 따로 처리하는 게 나을까요? 
아니면 지금 상태에서도 충분히 깔끔한 코드를 작성할 수 있을까요?

마음에 드는 좋은 방법이 안 떠오르네요.

오리.. 도와주세요...🤦 
<hr>

#### 2. 컨트롤러에서 재입력 로직 제거

자랑할 만한 변화는 아니지만, 재입력 로직이 지워져서 불편하실 거 같아 여기서 미리 말씀드립니다.
이번에 컨트롤러의 로직도 새로 작성하면서 재입력까지 넣으려니 잔에러도 계속 생기고 너무 복잡해지더라고요.

좀 더 깔끔하게 구현할 방법을 아직 못 찾아 일단 해당 로직을 제거하였습니다.
더 학습하고 다시 추가해 놓겠습니다.😿

jinyoungchoi95 commented on src/main/java/chess/domain/piece/Piece.java:
```diff
@@ -0,0 +1,33 @@
+            return type.findMoveStrategy(source, target) && source.isNotBackward(target, color);
+        }
+        return type.findMoveStrategy(source, target);
+    }
```
> Piece를 다시 추상 클래스로 만들고 폰만 Piece를 상속받게 해서 따로 처리하는 게 나을까요?
> 아니면 지금 상태에서도 충분히 깔끔한 코드를 작성할 수 있을까요?

저는 지금 만들어주신 추상화레벨만으로도 충분하다고 생각하고, `PAWN`에 대해서 if 분기를 타게되어 추상화레벨이 안맞는 것에 대해서도 공감해요. findMoveStrategy에서 color를 넘겨주어도 되지 않을까요? 혹은 `source`, `target`, `color`를 가진 객체를 findMoveStrategy(source+target+color)로 넘겨주어도 될 것 같구요.

color를 넘겨주는 것 자체가 싫으시다면, `BLACK_PAWN`, `WHITE_PAWN`을 별도로 분리하여 만드는것도 방법이겠네요. 각각 `color`를 알고 있으니 `isNotBackward`검증부를 사용할 수 있을테니까요.

정답은 없기때문에 냥인이 충분히 고민해보시고 좋은 방법으로 개선해주셔도 좋아요 :)

jinyoungchoi95 commented on src/main/java/chess/domain/square/File.java:
```diff
@@ -0,0 +1,67 @@
+    e(5),
+    f(6),
+    g(7),
+    h(8);
```
```suggestion
    h(8),
    ;
```
(반영하지 않으셔도 됩니다.) 추후 추가 편의성을 위해서 `,`를 뒤에 붙여두기도 해요.

cutehumanS2 replied:
추후 추가 편의성을 고려하지는 못 했네요! 반영하였습니다.😸

jinyoungchoi95 commented on src/main/java/chess/domain/square/File.java:
```diff
@@ -0,0 +1,67 @@
+
+    public boolean isSameFile(final File other) {
+        return index == other.index;
+    }
```
enum인데 index비교를 하지 않고 `==`을 외부에서 호출해도되지 않나요?

cutehumanS2 replied:
그렇네요.😄 해당 메서드를 삭제하고, `==`로 비교 수행하도록 변경하였습니다.

jinyoungchoi95 commented on src/main/java/chess/domain/square/File.java:
```diff
@@ -0,0 +1,67 @@
+        return filePath;
+    }
+
+    private File valueOf(final int value) {
```
```suggestion
    private File valueOfIndex(final int index) {
```
index를 파라미터로 받는다는 것을 명시하면 좀더 명확할 것 같아요 😄

cutehumanS2 replied:
반영하였습니다.😸

jinyoungchoi95 commented on src/main/java/chess/domain/square/Rank.java:
```diff
@@ -0,0 +1,78 @@
+        return (int) Math.signum(index - other.index);
+    }
+
+    public List<Rank> findRankPath(final Rank other) {
```
파라미터에 final을 붙인곳과 안붙인곳의 차이가 무엇일까요??

cutehumanS2 replied:
메서드 내에서 해당 파라미터의 재할당을 방지하기 위해 모두 붙여주도록 하고 있는데, 꼼꼼하지 못해서 몇몇 부분은 누락되었습니다.😅 모두 붙여주었습니다.

jinyoungchoi95 commented on src/main/java/chess/domain/piece/PieceColor.java:
```diff
@@ -0,0 +1,22 @@
+
+    public boolean isWhite() {
+        return this == WHITE;
+    }
```
```suggestion
```
현재 사용하지 않고 있는 것 같아요.

cutehumanS2 replied:
사용하지 않는 메서드들을 제거해 주었습니다!

jinyoungchoi95 commented on src/main/java/chess/domain/piece/PieceType.java:
```diff
@@ -0,0 +1,26 @@
+    KING((source, target) -> (source.isStraight(target) || source.isDiagonal(target))
+            && source.isWithinOneStep(target)),
+    QUEEN((source, target) -> source.isStraight(target) || source.isDiagonal(target)),
+    ROOK((source, target) -> source.isStraight(target)),
```
```suggestion
    ROOK(Square::isStraight),
```
(반영하지 않으셔도 됩니다.) 개인적으로 호불호가 많이 갈리는 문법이라고 생각하기 때문에 적용하지 않으셔도된다고 생각하는데요. 🤔 혹시나하여 코멘트만 남겨두어요.

cutehumanS2 replied:
저도 메서드 참조 방식을 더 선호합니다! 그러나 아래처럼 두 가지 이상의 메서드를 나열할 땐 적용이 안 돼서 통일성 때문에 일단 람다 표현식으로 두었는데, 이 정도까지의 통일성은 안 줘도 될까요?

```java
QUEEN((source, target) -> source.isStraight(target) || source.isDiagonal(target)),
```

jinyoungchoi95 replied:
반영하지 않으셔도 된다고 말씀드렸던 이유가 개인의 취향이라 저는 통일성을 줘도 괜찮고 안줘도 괜찮다고 생각합니다 😄

jinyoungchoi95 commented on src/main/java/chess/domain/piece/PieceType.java:
```diff
@@ -0,0 +1,26 @@
+
+import java.util.function.BiPredicate;
+
+public enum PieceType {
```
`PieceType`에서 각각의 기물에 대한 source to target 가능 여부를 판단하도록 바꾸어주셨네요 👍 

moveStrategy로 전략으로 생각하시는 부분이 저도 좋다라고 생각하는데요. `PieceType` enum이 아닌 `PieceMoveStrategy` interface로 추상화하고 `BiPredicate<Square, Square> moveStrategy;`의 각 로직을 구현체로 두는 건 어떨까요?

현재의 구조도 좋다라고 생각하지만 predicate로는 로직이 다소 복잡해보이고 동일한 로직을 interface로 추상화해도 좋을 것 같아보여서요. 재사용이 걱정이라면 전략 구현체들을 싱글톤으로 만들수도 있겠네요.

cutehumanS2 replied:
Predicate를 이용하여 이동 전략들을 나열했을 때 복잡해 보인다는 것에 동의하여, 
각 기물의 이동 전략을 인터페이스로 추상화하였습니다.

여기서 고민은 전략들을 어떻게 세분화 할지 입니다.

1. 이동 방법에 따라 나눈다.
`StraightStrategy`, `DiagonalStrategy`, `WithinOneStepStrategy`, `OnlyForwardStrategy`, `PawnAttackStrategy` 로 구현체를 만들고 조합하는 방식입니다.<br>
 기존 `||` 조건은 `UnionStrategy` 를, `&&` 조건은 `IntersectionStrategy` 를 사용하여 조합할 수 있습니다(전략명은 마땅히 떠오르는 게 없어 `||`와 `&&`의 의미 그대로 지었습니다😅).<br>
`StraightStrategy`와 `DiagonalStrategy` 가 여러 번 사용될 수 있다는 걸 염두에 두고 이런 방식을 떠올렸는데요, 아래 보이는 것과 같이 가독성이 좋지 않은 것 같아요. 이전에 Predicate로 조건을 나열했을 때랑 비슷한 수준으로 복잡해 보입니다.
    ```java
     KING(new IntersectionStrategy(List.of(
            new UnionStrategy(List.of(new StraightStrategy(), new 
     DiagonalStrategy())),
            new WithinOneStepStrategy()))),
     QUEEN(new UnionStrategy(List.of(new StraightStrategy(), new 
     DiagonalStrategy()))),
     ...
     KNIGHT(new KnightStrategy()),
     ...
     ``` 
     게다가 Knight의 경우 `StraightAndDiagonalStrategy`라 짓기엔 `Intersection(Straight, Diagonal)`과 혼동될 거 같아 `KnightStrategy`라 네이밍 했는데, 이 때문에 통일성이 깨지는 것은 물론 ‘이동 방법에 따라 나눈다’라는 조건에도 맞지 않습니다.

2) 기물에 따라 나눈다.

    그래서 차라리 `KnightStrategy`, `KingStrategy` , … 식으로 기물에 따라 전략을 나누면 어떨까 싶기도 했는데, 이 또한 완전히 만족스럽지는 않습니다.😅

    전략 패턴이 객체가 할 수 있는 행위를 전략 클래스로 구현해서 상호교환이 가능하도록 만드는 것이라 생각하는데, 기물마다 나눈다면 과연 전략 패턴의 특징과 장점을 살린 게 맞을까 싶어요.

    구현은 2번이 좀 더 깔끔할 것 같긴 합니다만, 섣불리 바꾸지는 못했습니다.

<br>

현재는 1번의 방식을 따르고 있습니다. 어떤 방식이 더 나을까요?
괜히 1번 방식으로 구현해서 더 어려운 길을 걷고 있는 게 아닌가 싶기도 합니다.

+) 그리고 PieceTypeTest, SquareTest, 각 StrategyTest의 내용이 중복되는데, 원래 중복되는 테스트여도 작성하는 게 맞나요?

질문이 너무 길고 복잡하죠.. 이것저것 판단이 안 서는 게 많아 자꾸 질문만 드리게 되네요.😹

jinyoungchoi95 replied:
정답은 없는 문제이기 때문에 1이어도 좋고 2여도 괜찮을 것 같아요. 2에서 말씀하신대로 기물에 대한 전략이 명확하기 때문에 구현했을 때 복잡도가 낮아 2가 더 좋아보이긴하네요.

중복 테스트의 경우 TDD를 했다면 기능명세를 먼저하고, 테스트를 작성 후 코드를 짰기 때문에 중복되어도 모든 경우에 대한 테스트검증이 존재하는게 맞지 않을까요?

jinyoungchoi95 commented on src/main/java/chess/domain/piece/PieceType.java:
```diff
@@ -0,0 +1,26 @@
+
+    public boolean findMoveStrategy(final Square source, final Square target) {
+        return moveStrategy.test(source, target);
+    }
```
반환 타입은 boolean이지만 `find`는 "찾다"의 의미이기 때문에 네이밍이 수정되면 좋겠네요.

cutehumanS2 replied:
source에서 target으로 이동할 수 있는지 여부를 판단해주는 메서드이므로 `canMove()`로 변경하였습니다.

jinyoungchoi95 commented on src/main/java/chess/domain/piece/PieceType.java:
```diff
@@ -0,0 +1,26 @@
+        this.moveStrategy = moveStrategy;
+    }
+
+    public boolean findMoveStrategy(final Square source, final Square target) {
```
각 전략에 대한 테스트가 필요해보여요.

jinyoungchoi95 commented on src/main/java/chess/domain/piece/Piece.java:
```diff
@@ -0,0 +1,33 @@
+
+import chess.domain.square.Square;
+
+public class Piece {
```
조합으로 잘 변경해주셨네요 👍

jinyoungchoi95 commented on src/main/java/chess/domain/square/Square.java:
```diff
@@ -0,0 +1,99 @@
+
+    public boolean isStraightAndDiagonal(final Square target) {
+        if (file.calculateDistance(target.file) == ONE_STEP
+                && rank.calculateDistance(target.rank) == TWO_STEP) {
```
if문안에 &&절도 private메소드로 분리하면 가독성이 더 좋을 것 같아요 :)

cutehumanS2 replied:
반영하였습니다!

jinyoungchoi95 commented on src/main/java/chess/domain/square/Square.java:
```diff
@@ -0,0 +1,99 @@
+    }
+
+    private boolean canAttack(final Square target) {
+        return target != null;
```
null을 활용한 로직들은 모두 제거해주세요. null은 어떤 값으로 인해 문제가 발생했는지 의미를 파악하기 힘들 뿐더러 다양한 문제를 발생시킵니다.
https://www.mimul.com/blog/why-null-is-bad/

cutehumanS2 replied:
제공해 주신 포스팅의 내용을 참고하여 null 대신 null object를 적용해 보았는데요, 
그 결과 Board에 폰의 공격 가능 여부를 판단하기 위한 조건문들이 추가되었습니다. 
잘못된 방향인 것 같아 더 좋은 방법을 찾고 싶었는데 해결이 안 되어 일단 이 상태로 두었습니다.😹 이쯤 되니 현재 제가 작성한 구조가 잘못된 게 아닌가 싶은데, 오리가 보기에는 어떠신가요?
```java
    private void validateCannotMove(final Square source, final Square target) {
        final Piece sourcePiece = findPieceBySquare(source);
        if (findPieceBySquare(source).isPawn()) {
            if (!canPawnAttack(source, target)) {
                throw new IllegalArgumentException(ERROR_MOVE_NOT_AVAILABLE);
            }
        }

        if (!sourcePiece.canMove(source, target)) {
            throw new IllegalArgumentException(ERROR_MOVE_NOT_AVAILABLE);
        }
    }

    private boolean canPawnAttack(final Square source, final Square target) {
        if (source.isDiagonal(target) && findPieceBySquare(target).isEmpty()) {
            return false;
        }
        return true;
    }
```

jinyoungchoi95 replied:
각 기물이 공격가능여부를 찾고 있는 현재의 구조에서 `Pawn`만을 위해 `Board`에 조건이 추가되었다면 그냥 startegy를 넘길 때 source, target과 함께 Board의 정보를 함께 넘겨주면 되지 않을까요?

jinyoungchoi95 commented on src/test/java/chess/domain/square/SquareTest.java:
```diff
@@ -0,0 +1,186 @@
+import java.util.List;
+
+@DisplayName("좌표")
+class SquareTest {
```
케이스가 많아지면서 무엇을 테스트하고자하는지 가독성이 떨어질수도 있는데요. `@Nested`를 메소드별로 선언하여 적절하게 사용하면 무슨 메소드를 테스트하려고하는지 계층이 분리되어 가독성에 도움을 주어요.

cutehumanS2 replied:
반영하였습니다. 
`@Nested` 어노테이션을 사용하니 확실히 테스트의 가독성이 향상된 것 같습니다.😸

jinyoungchoi95 commented on src/main/java/chess/domain/square/Square.java:
```diff
@@ -0,0 +1,99 @@
+
+    public boolean isOnlyForward(final Square target) {
+        if (isFirstMove()) {
+            return rank.calculateDistance(target.rank) <= TWO_STEP
+                    && file.isSameFile(target.file);
+        }
+        return rank.calculateDistance(target.rank) == ONE_STEP
+                && file.isSameFile(target.file);
+    }
+
+    private boolean isFirstMove() {
+        return rank.isSameRank(Rank.TWO) || rank.isSameRank(Rank.SEVEN);
+    }
+
+    public boolean isAttack(final Square target) {
+        if (canAttack(target)) {
+            return isDiagonal(target)
+                    && file.calculateDistance(target.file) == 1;
```
```suggestion
            return rank.calculateDistance(target.rank) <= TWO_STEP && file.isSameFile(target.file);
        }
        return rank.calculateDistance(target.rank) == ONE_STEP && file.isSameFile(target.file);
    }

    private boolean isFirstMove() {
        return rank.isSameRank(Rank.TWO) || rank.isSameRank(Rank.SEVEN);
    }

    public boolean isAttack(final Square target) {
        if (canAttack(target)) {
            return isDiagonal(target) && file.calculateDistance(target.file) == 1;
```
내려간 이유가 있을까요? 제기억에 우테코에서 제공된 컨벤션으로는 이정도 길이면 인텔리제이 리펙터링에서 안내려가는 길이로 알고 있어서요. (아니라면 넘어가주세요 😄 )

cutehumanS2 replied:
제가 보기에 가독성이 안 좋은 경우 한 칸 내렸습니다.😸
컨벤션에 맞게 다시 한 줄로 고쳤습니다! ㅎㅎㅎ

jinyoungchoi95 commented on src/main/java/chess/domain/board/BoardFactory.java:
```diff
@@ -0,0 +1,61 @@
+
+public class BoardFactory {
+
+    public static final List<Function<PieceColor, Piece>> PIECES_ARRANGEMENT = List.of(
```
```suggestion
    private static final List<Function<PieceColor, Piece>> PIECES_ARRANGEMENT = List.of(
```
👀

jinyoungchoi95 commented on src/test/java/chess/domain/board/BoardFactoryTest.java:
```diff
@@ -0,0 +1,64 @@
+    void createBoard() {
+        final Board actual = BoardFactory.createBoard();
+
+        final Map<Square, Piece> expected = Map.ofEntries(
```
대용량 테스트군요 👍

jinyoungchoi95 commented on src/main/java/chess/domain/board/Board.java:
```diff
@@ -0,0 +1,80 @@
+        validateIsTurn(sourcePiece, turn);
+        validateCanMove(source, target, sourcePiece);
+        validateExistObstacleOnPath(source, target);
+        pieces.remove(source);
+        pieces.put(target, sourcePiece);
```
둘은 하나의 메소드로 묶어 `replace~`로 표현될수도 있겠네요 :)

cutehumanS2 replied:
`replacePieceSquare()` 메서드로 분리해 보았습니다.😄

jinyoungchoi95 commented on src/main/java/chess/domain/board/Board.java:
```diff
@@ -0,0 +1,80 @@
+    }
+
+    public void move(final Square source, final Square target, final PieceColor turn) {
+        validateIsSameSquare(source, target);
+        validateIsNonExistentPiece(source);
+
+        Piece sourcePiece = findPieceBySquare(source);
+        validateIsTurn(sourcePiece, turn);
+        validateCanMove(source, target, sourcePiece);
+        validateExistObstacleOnPath(source, target);
```
`validateIsSameSquare`는 출발지와 목적지가 같으면 예외가 발생하며 `isSameSquare` 메소드 명의 조건이면 예외가 발생해요. 다만 `validateIsTurn`은 반대로 현재턴이 아닌 경우에 예외가 발생하며 `isTurn` 메소드 명의 반대조건이면 예외가 발생해요. 

validation 메소드명이 통일성있으면 좋을 것 같아요 :)

cutehumanS2 replied:
`validateIsNotTurn()`으로 변경하였습니다. 
다른 메서드도 통일성있게 변경하였습니다!

jinyoungchoi95 commented on src/test/java/chess/domain/board/BoardTest.java:
```diff
@@ -0,0 +1,100 @@
+import java.util.Map;
+
+@DisplayName("보드")
+class BoardTest {
```
테스트 👍

jinyoungchoi95 reviewed (changes requested):
안녕하세요 냥인 😄

피드백 잘 반영해주셨네요. 몇가지 코멘트남겨두었으니 확인부탁드려요.

궁금하거나 고민이 되는 부분이 있으시다면 언제든 pr 코멘트 또는 dm으로 요청 부탁드립니다.
감사합니다 🙇‍♂️

cutehumanS2 committed:
style: 추후 추가 편의성을 고려하여 Enum 클래스 코드 스타일 변경

cutehumanS2 committed:
refactor: 메서드 및 파라미터 네이밍 변경

+) CommandTest에 DisplayName 추가

cutehumanS2 committed:
refactor: final 키워드로 재할당 방지

cutehumanS2 committed:
refactor: 불필요한 메서드 제거

cutehumanS2 committed:
refactor: 각 기물의 이동 전략을 인터페이스로 추상화

cutehumanS2 committed:
refactor: 검증 메서드명 변경

cutehumanS2 committed:
refactor: Piece의 Square를 교체하는 로직을 메서드화

cutehumanS2 committed:
refactor: 접근 제어자 변경

cutehumanS2 committed:
refactor: 테스트 보완

- @Nested 어노테이션 사용
- final 키워드 사용
- 테스트 데이터 추가

cutehumanS2 committed:
refactor: null 대신 null object 적용

cutehumanS2 committed:
refactor: 조건문을 메서드로 분리

cutehumanS2 committed:
refactor: 불필요한 코드 제거

cutehumanS2 committed:
test: PieceType 테스트

jinyoungchoi95 commented on src/main/java/chess/domain/piece/strategy/IntersectionStrategy.java:
```diff
@@ -0,0 +1,24 @@
+    @Override
+    public boolean canMove(final Square source, final Square target) {
+        for (final MoveStrategy moveStrategy : moveStrategies) {
+            if (!moveStrategy.canMove(source, target)) {
+                return false;
+            }
```
indent가 2가 넘어가네요.

jinyoungchoi95 commented on src/main/java/chess/view/OutputView.java:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    public static void printBoard(final Map<Square, Piece> board) {
+        final char[][] result = generateEmptyBoard();
```
> 배열 대신 컬렉션을 사용한다.

jinyoungchoi95 reviewed (approved):
안녕하세요 냥인 😄

피드백 잘 반영해주셨네요. 다음 미션 진행을 위해 머지하겠습니다.

궁금하거나 고민이 되는 부분이 있으시다면 언제든 pr 코멘트 또는 dm으로 요청 부탁드립니다.
감사합니다 🙇‍♂️
